machine/src/chip.rs

```diff
@@ -22,31 +20,286 @@ pub trait Chip<M: Machine> {
         vec![]
     }
 
-    fn global_sends(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_sends(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
 
-    fn global_receives(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_receives(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
+
+    fn all_interactions(&self, machine: &M) -> Vec<(Interaction<M::F>, InteractionType)> {
+        let mut interactions: Vec<(Interaction<M::F>, InteractionType)> = vec![];
+        interactions.extend(
+            self.local_sends()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalSend)),
+        );
+        interactions.extend(
+            self.local_receives()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalReceive)),
+        );
+        interactions.extend(
+            self.global_sends(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalSend)),
+        );
+        interactions.extend(
+            self.global_receives(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalReceive)),
+        );
+        interactions
+    }
+
+    fn interaction_map(&self, machine: &M) -> BTreeMap<BusArgument, Vec<usize>> {
+        let mut map: BTreeMap<BusArgument, Vec<usize>> = BTreeMap::new();
+        for (n, (interaction, _)) in self.all_interactions(machine).iter().enumerate() {
+            map.entry(interaction.argument_index)
+                .or_insert_with(Vec::new)
+                .push(n);
+        }
+        map
+    }
+}
+
+pub trait ValidaAirBuilder: PermutationAirBuilder {
+    type Machine;
+
+    fn machine(&self) -> &Self::Machine;
+
+    fn cumulative_sums(&self) -> &[Self::EF];
 }
 
 pub struct Interaction<F: Field> {
     pub fields: Vec<VirtualPairCol<F>>,
     pub count: VirtualPairCol<F>,
-    pub argument_index: usize,
+    pub argument_index: BusArgument,
 }
 
-/// Generate the permutation trace for the chip given the provided machine.
-fn generate_permutation_trace<M: Machine, C: Chip<M>, EF: ExtensionField<M::F>>(
-    chip: &C,
+#[derive(Clone)]
+pub enum InteractionType {
+    LocalSend,
+    LocalReceive,
+    GlobalSend,
+    GlobalReceive,
+}
+
+#[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub enum BusArgument {
+    Local(usize),
+    Global(usize),
+}
+
+impl<F: Field> Interaction<F> {
+    pub fn is_local(&self) -> bool {
+        match self.argument_index {
+            BusArgument::Local(_) => true,
+            BusArgument::Global(_) => false,
+        }
+    }
+
+    pub fn is_global(&self) -> bool {
+        match self.argument_index {
+            BusArgument::Local(_) => false,
+            BusArgument::Global(_) => true,
+        }
+    }
+
+    pub fn argument_index(&self) -> usize {
```

Comment:
Is this unfinished?

Reply:
What do you mean?

Reply:
Oh sorry, I misinterpreted this as giving "universal" indices